File: Assuntos da Prova 2/Romberg.c
#include <math.h>
#include <stdio.h>

#define error_order 32
// numElemsFirstCol deve ser error_order / 2
#define numElemsFirstCol error_order / 2

// Método da regra dos trapézios
double trapz(double (*f)(double), double a, double b, int n) {
    double soma = 0;
    double h = (double)(b - a) / (double)n;
    for (int k = 1; k < n; k++) {
        soma += f(a + k * h);
    }
    soma *= 2;
    soma += f(a);
    soma += f(b);
    soma *= (h / 2.0);
    return soma;
}

void romberg(double array[]) {
    // i=0 está calculando a coluna f2
    int numCols = numElemsFirstCol - 1;
    for (int i = 0; i < numCols; i++) {
        for (int j = 0; j < numCols; j++) {
            double numer = pow(2, (i + 1) * 2) * array[j + 1] - array[j];
            double denom = pow(2, (i + 1) * 2) - 1;
            array[i] = numer / denom;
        }
    }

    printf("Aprox O(h^%d) = %.16f\n", error_order, array[0]);
}

double f(double x) {
    return exp(-x * x);
}

int main(int argc, char const *argv[]) {
    // Exemplo:
    // Aproximar int exp(-x*x), de 0 a 1
    double a = 0, b = 1, h = 0.5;

    int n = (int)((b - a) / h);

    printf("n = %.16f\n\n", (double)n);
    printf("Error_order: %d \nnumElemensFirstCol: %d\n\n", error_order, numElemsFirstCol);

    double coluna_F1[numElemsFirstCol];
    for (int i = 0; i < numElemsFirstCol; i++) {
        coluna_F1[i] = trapz(f, a, b, (i + 1) * n);
    }
    romberg(coluna_F1);

    return 0;
}

File: Assuntos_P2/Algoritmos/romberg.py
import math

def romberg(col_1):
    n = len(col_1)
    for i in range(n-1):
        for j in range(n-1-i):
            numer = 4 ** (i+1) * col_1[j + 1] - col_1[j]
            denom = 4 ** (i+1) - 1
            value = numer / denom
            col_1[j] = value
    return col_1[0]


def trapz(f, a, b, h):
    n = int((b-a) / h)
    soma = 0
    for k in range(1, n):
        soma += f(a + k * h)
    return (h/2) * (f(a) + 2* soma + f(b))


def f(x):
    return math.exp(-x**2)


if __name__ == '__main__':
    a = 0
    b = 1

    h = 0.5
    k = 5
    hs = [h/2 ** i for i in range(k)]
    
    col1 = [trapz(f, a, b, hi) for hi in hs]

    r = romberg(col1)

    print(f'approx = {r}')